packages/core/graph/src/AdjacencyList.js

```diff
@@ -0,0 +1,1195 @@
+ *       │  │  │  │  │  │  │  ...  │  │  │  │  │
+ *       └──┴──┴──┴──┴──┴──┴───────┴──┴──┴──┴──┘
+ *       └──┬──┘                   └─────┬─────┘
+ *        header                        node
+ * (NODES_HEADER_SIZE)              (NODE_SIZE)
+ *
+ * The header for the nodes array comprises 2 4-byte chunks:
+ * The first 4 bytes store the node capacity.
+ * The second 4 bytes store the number of nodes in the adjacency list.
+
+ * struct NodesHeader {
+ *   int capacity;
+ *   int count;
+ * }
+ *
+ * ┌────────────────────────┐
+ * │    NODES_HEADER_SIZE   │
+ * ├────────────┬───────────┤
+ * │  CAPACITY  │   COUNT   │
+ * └────────────┴───────────┘
+ *
+ * Each node is represented with 4 4-byte chunks:
+ * The first 4 bytes are the hash of the node's first incoming edge.
+ * The second 4 bytes are the hash of the node's first outgoing edge.
+ * The second 4 bytes are the hash of the node's last incoming edge.
+ * The second 4 bytes are the hash of the node's last outgoing edge.
+ *
+ * struct Node {
+ *   int firstIn;
+ *   int firstOut;
+ *   int lastIn;
+ *   int lastOut;
+ * }
+ *
+ * ┌─────────────────────────────────────────────────┐
+ * │                    NODE_SIZE                    │
+ * ├────────────┬───────────┬───────────┬────────────┤
+ * │  FIRST_IN  │ FIRST_OUT │  LAST_IN  │  LAST_OUT  │
+ * └────────────┴───────────┘───────────┴────────────┘
+ */
+export const NODE_SIZE: 4 = 4;
+/** The size of nodes array header */
+export const NODES_HEADER_SIZE: 2 = 2;
+
+/**
+ * Edges are stored in a shared array buffer of fixed length
+ * equal to the edge `capacity + capacity * EDGE_SIZE + EDGES_HEADER_SIZE`.
+ *
+ *                  hash table                                   edge
+ *                  (capacity)                               (EDGE_SIZE)
+ *               ┌──────┴──────┐                         ┌────────┴────────┐
+ *      ┌──┬──┬──┬──┬───────┬──┬──┬──┬──┬──┬──┬──┬───────┬──┬──┬──┬──┬──┬──┐
+ *      │  │  │  │  │  ...  │  │  │  │  │  │  │  │  ...  │  │  │  │  │  │  │
+ *      └──┴──┴──┴──┴───────┴──┴──┴──┴──┴──┴──┴──┴───────┴──┴──┴──┴──┴──┴──┘
+ *      └───┬────┘             ├─────────────────────┬─────────────────────┘
+ *        header        addressableLimit           edges
+ * (EDGES_HEADER_SIZE)               (capacity * EDGE_SIZE * BUCKET_SIZE)
+ *
+ * The header for the edges array comprises 3 4-byte chunks:
+ * The first 4 bytes store the edge capacity.
+ * The second 4 bytes store the number of edges in the adjacency list.
+ * The third 4 bytes store the number of deleted edges.
+ *
+ * struct NodesHeader {
+ *   int capacity;
+ *   int count;
+ *   int deletes;
+ * }
+ *
+ * ┌────────────────────────────────────┐
+ * │          EDGES_HEADER_SIZE         │
+ * ├────────────┬───────────┬───────────┤
+ * │  CAPACITY  │   COUNT   │  DELETES  │
+ * └────────────┴───────────┴───────────┘
+ *
+ * Each edge is represented with 8 4-byte chunks:
+ * The first 4 bytes are the edge type.
```

Comment:
We probably don't need to allocate that much space for the type? 1 byte (max 256 types) would be more than enough. Though I guess that makes things a little more complex and you'd have to deal with alignment...

Reply:
This is a really good point, and something I expanded on a bit in [this comment](https://github.com/parcel-bundler/parcel/pull/6922#discussion_r718652525).